[1~2단계 - 방탈출 예약 대기] 헤일러(민서현) 미션 제출합니다

threepebbles opened the pull request:
안녕하세요 파랑! 헤일러입니다. 리뷰이로 만나뵙게 되어 반갑습니다 😄 

저는 1년 전쯤 김영한님의 JPA 강의를 수강했던 기억이 있었고, 페어는 JPA를 프로젝트에 사용해본 경험이 있어 JdbcTemplate를 JPA로 전환하는 것은 크게 어렵지는 않았던 것 같습니다.

![image](https://github.com/user-attachments/assets/1994b7d6-74a2-46a3-938b-e4bf48f77cea)

이전에 인터넷 강의만 들으며 학습할 당시에는 JdbcTemplate의 존재를 모르고 바로 JPA만 사용했었기 때문에 JPA의 장점을 크게 느끼지 못했었는데, 지금은 JPA가 많은 것을 자동화해주고 있었다는 것을 몸소 느끼고 있습니다.

이번 미션에 JPA를 사용하며 느낀 가장 큰 장점은 `@Entity` 어노테이션이 붙은 클래스로 생성된 객체를 사용할 때 데이버테이스 테이블, SQL 쿼리문을 거의 신경 쓰지 않을 수 있었다는 것입니다.

클래스를 테이블, 객체를 테이블의 한 행, 객체 간의 참조 관계를 외래 키로 자동 매핑해주는 그 일련의 과정을 자동화해줌으로써, 데이터베이스 테이블에 의존적인 제약 조건과 기능을 크게 신경쓰지 않을 수 있었습니다. 
조금 극단적으로 말하면 간단한 CRUD 기능을 구현한다 했을 때 개발자 입장에서는 데이터베이스의 존재조차 몰라도 되고, 그냥 도메인 객체를 어느 저장소(데이터베이스든 CSV파일이든 어딘가)에 저장하고 조회해온다만 인지해도 사용할 수 있을 정도란 생각이 들었습니다.

한편으로는 자동화해주는 부분이 많은 만큼 특정 부분(ex. 복잡한 저장/조회 로직)에서는 개발자가 작성한 소스 코드만으로는 전혀 예측할 수 없는 방식으로 동작하는 경우도 꽤 있을 것이란 생각도 들었습니다.

그래서 JPA, JPA를 구현해놓은 Hibernate가 개발자의 어떤 불편함을 어떤 방식으로 해소하려고 했는지 그 철학과, 예측하기 어려운 자동화된 내부 동작에 대해 어느정도 이해하고 조심하면서 사용해야 할 필요성을 느끼고 있습니다.

그런 관점에서 사용하는 연관관계 매핑에 따라 의도치 않은 중간테이블이 생성되는지, 의도한대로 실제 쿼리가 발생하는지, 키 생성 전략(IDENTITY, SEQUENCE)에 따라 flush()가 일어나는 시점이 달라질 수 있는지 그런 점들을 짚어가면서 학습하고 있습니다.

리뷰 잘 부탁드립니다! 😁 

## 아래는 PR 리뷰 요청 범위입니다.

[미션3 Files Changed](https://github.com/woowacourse/spring-roomescape-waiting/pull/177/files/1d9f9d12d99c10a21d525c27000b175a3ddc4c59..29f862f8549bf22dd8fc85896043fb9ace03c455)

---


## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [헤일러의 프롤로그](https://prolog.techcourse.co.kr/studylogs/4335)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
✅: 잘 지킴, ❓: 애매함, 잘 못 지킴
규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다. ✅
규칙 2: else 예약어를 쓰지 않는다. ✅
규칙 3: 모든 원시값과 문자열을 포장한다. ❓
규칙 4: 한 줄에 점을 하나만 찍는다. ✅
규칙 5: 줄여쓰지 않는다(축약 금지). ✅
규칙 6: 모든 엔티티를 작게 유지한다. ✅
규칙 7: 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다. ❓
규칙 8: 일급 콜렉션을 쓴다. ❓
규칙 9: 게터/세터/프로퍼티를 쓰지 않는다. ❓

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?

기능 구현을 꼼꼼하게 하면서 남는 시간에는 어떻게 하면 신뢰 높은 테스트를 작성할 수 있을지 고민하고 있습니다.

현재 작성한 테스트에 부족한 점(ex. 테스트 픽스처의 올바른 사용, e2e 테스트할 때 유의해야 할 점 등), 학습해보면 좋을만한 키워드가 있다면 추천해주시면 감사하겠습니다. 😄 

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

woowabrie committed:
feat: setup project

boorownie committed:
chore: upgrade build environment (Gradle 8.13, JDK 21, Spring Boot 3.4.4)

threepebbles committed:
Mission3 Code Base Migration

[1~3단계 - 방탈출 사용자 예약] 헤일러(민서현) 미션 제출합니다.  (#195)

* docs: add PR template

* migrate from mission 1

* feat : 예약 시간 생성 API 응답 상태 코드 수정

- 200 -> 201

* feat : 예약 생성 API 응답 상태 코드 수정

- 200 -> 201

* refactor : 응답 메시지 시간 형식 지정

* refactor : 요청 메시지 시간 형식 지정

* docs : 발생할 수 있는 예외 상황 추가

* feat : 예약 시간 범위 검증 추가

* feat : 예약시간 중복 검증 추가

* feat : 삭제하려고 하는 예약 시간을 사용하고 있는 예약 정보가 있는지 검증하는 기능 추가

* chore : 테마 테이블 스키마 추가

* feat : Domain Entity와 Database Entity 분리

* chore : 테마 테이블명 변경

- theme -> themes

* feat : 테마 Domain Entity, Database Entity 추가

* feat : 테마 레포지토리 추가

* feat : 테마 관리 API 추가

- 테마 조회, 추가, 삭제 API

* refactor : 메서드명 수정

* feat : 참조 중인 예약 정보 행을 삭제 요청시 발생하는 예외 처리 수정

* refactor : 메서드명 수정

* docs : 테마 요구사항 추가

* feat : 중복된 테마명 검증 기능 추가

* chore : URL 슬래시 추가

* feat : 한 테마에 대해서 같은 날짜와 시간에 예약된 정보가 있는지 검증 추가

* docs : 사용자 예약 화면 응답 추가

* feat : 사용자 예약 화면 응답 기능 추가

* chore : 생성자 검증 로직 리팩터링

* chore : 테마명에 UNIQUE 제약 조건 추가

* feat : 날짜와 테마명에 따른 예약 가능한 시간 검증 기능 추가

* feat : 예약 시간 제거 API 응답 상태 코드 수정

* chore : 불필요한 공백 제거

* chore : final 키워드 추가

* refactor : Domain Entity, Database Entity 병합

* docs : 유저 예약 시스템 추가

* feat : 홈 화면 API 응답 추가

* feat : 일주일 이내 가장 예약이 많았던 인기 테마 10개를 응답하는 API 추가

* chore : 메서드 선언 순서 변경

* test : 예약 서비스 테스트 추가

* test : 예약 생성자 검증 테스트 추가

* test : 날짜와 시간이 중복되는 테마 검증 테스트 추가

* test : 예약 시간 테스트 추가

* test : 테마 서비스 테스트 추가

* test : 테마 테스트 추가

* chore : 공백 정리

* test : 예약 시간 컨트롤러 API 테스트 추가

* test : 테마 컨트롤러 API 테스트 추가

* test : TestConfiguration 추가

* feat : 예약 정보 삭제 API 응답 상태 코드 수정

- ok -> no content

* test : 예약 컨트롤러 API 테스트 추가

* chore : 공백 정리

* chore : 변수명 오타 수정

* chore : 패키지 이동

* refactor : SQL Query 개선

* remove : 테스트 schema.sql 파일 삭제

* chore: 중복된 의존성 제거

* refactor: 커스텀 예외 클래스 네이밍 개선

* refactor: 레포지토리명 개선

- H2ReservationRepository -> JdbcReservationRepository
- H2ReservationTimeRepository -> JdbcReservationTimeRepository

* refactor: ControllerAdvice -> RestControllerAdvice 수정

* refactor: 어노테이션 순서 배치를 일관되게 수정

- Spring Stereotype 어노테이션, 생성자/필드 관련 어노테이션, 핸들러 매핑 관련 어노테이션 순서로 정리

* refactor: 가독성 개선을 위해 메서드 체이닝 중간에 개행 추가

* docs: 기능 설명 추가

* refactor: view 이름 반환 시 불필요한 '/' 제거

* test: 빈 문자열 검증 테스트 추가

* test: 테마 생성, 삭제 예외 테스트 추가

* test: 변수명 변경

* chore: 예약 시간 스키마 변경

- start_at 타입: VARCHAR(255) -> TIME

* feat: 유저 예약 시스템에서 특정 날짜, 테마 id로 예약 가능한 시간을 조회하는 API의 HTTP 메서드 수정(POST -> GET)

* test: 예약 가능한 시간 목록 조회 API 테스트 개선

* refactor: toString 재정의에 Lombok 기능 사용

---------

Co-authored-by: solarsalt <solar@woowahan.com>
Co-authored-by: bowook <mbosx316@naver.com>

refactor: 패키지 구조 대규모 변경

- DDD 기반 패키지 구조로 변경(application, domain, infrastructure, ui)

chore(build.gradle): JWT 관련 기능 의존성 추가

refactor(theme 패키지): service 메서드가 controller의 핸들러 메서드에 일대일 대응 되도록 수정

refactor(reservation 패키지): service 메서드가 controller의 핸들러 메서드 기능에 일대일 대응되도록 수정

refactor: 테스트 패키지 구조를 프로덕션 패키지 구조와 동일하게 변경

chore: members 테이블 추가

chore(data.sql): 초기 데이터 추가

chore(build.gradle): jjwt 라이브러리 의존성 수정

fix(Reservation): 예약 테이블 스키마 변경에 따라 예약 도메인 객체 수정

feat(UserRole): 권한 구분을 위한 enum 클래스 추가

feat(RequiresRole): Controller의 HandlerMethod 별 권한 구분을 위한 커스텀 어노테이션 추가

feat(Member): 멤버 추가, 삭제 기능 추가

fix: 예약 테이블 스키마 변화로 인한 에러 수정

feat(프론트엔드): 요구사항에 맞게 html, js 수정

feat: JWT를 이용한 로그인, 로그아웃 기능 추가

test: 예약 테이블 스키마 변경으로 인해 통과하지 않는 테스트 주석 처리

refactor: 클래스명 좀 더 명확하게 개선, AuthRole 패키지 이동

- UserRole -> AuthRole

feat(CheckUserRoleInterceptor): HandlerMethod에 RequiresRole 어노테이션이 없으면 모두에게 열린 public API로 취급하도록 수정

refactor: UserRole -> AuthRole 네이밍 수정

fix(CheckAuthRoleInterceptor): public API는 @RequiredRole 어노테이션을 사용하지 않아도 동작하도록 수정

fix(MemberArgumentResolver): token 유효성 검증 로직 추가

fix(JwtTokenProvider): JWT payload에 AuthRole의 roleName 정보를 담도록 수정

fix: 예약 테이블 조회 쿼리 문법 오류 수정

fix(프론트엔드): 응답 상태코드에 따른 처리가 잘못된 부분 수정

- 예약 성공 시 ok가 아닌 created가 응답됨

feat: 예약 내역 조회 응답에 예약자 이름도 포함하도록 수정

fix(프론트엔드): 4단계, 5단계에 활성화되어야 하는 기능 주석 해제

feat: admin용 예약 내역 조회 엔드포인트 분리

remove: 사용하지 않는 html 삭제

chore(프론트엔드): 쿼리 파라미터 변수 네이밍 개선

feat: 전체 Member의 이름 조회 기능 추가

feat: admin 페이지에서 조건 기반 예약 내역 조회 기능 추가

- 엔드포인트: GET /admin/reservations

feat: 로그아웃 로직 수정

refactor: 패키지 이동

- View Resolver를 통해 서버 사이드 렌더링하는 Controller 분리

refactor: controller에서 응답 dto를 재사용하도록 수정

fix(프론트엔드): 예약 내역 응답 데이터 파싱 로직 수정

fix: Member 목록 조회 요청 시 Member의 id도 응답하도록 수정

refactor: 함수 선언 순서 변경

refactor: 핸들러 메서드 더 적절한 클래스로 이동

docs: API 엔드포인트, 예외 처리 기준으로 문서 작성

refactor: 개수를 세는 것이 아닌 존재 여부만 확인하도록 쿼리 수정

- count -> exists

refactor: DB 호환성을 위해 쿼리문에서 날짜를 LocalDate 변수로 추출, 레포지토리명 개선

- 클래스명 개선: H2 -> Jdbc

chore(data.sql): 개발 단계에서 브라우저 테스트를 쉽게 하기 위한 초기 데이터 추가

chore: email 컬럼에 UNIQUE 제약 조건 추가

refactor: 조회된 Member의 상태를 사용하도록 수정

chore: 예외 메시지 개선

chore: application.yml 주석 추가

chore: 테스트용 data.sql 추가

chore: 엔드포인트 권한으로 인해 실패하는 테스트 주석 처리

chore: 테스트 TODO 추가

remove: 사용하지 않는 클래스 제거

test: ReservationService 테스트 복원, 테스트 픽스처 추가

- ReservationTime, Theme, Member 테스트 픽스처 추가

feat: 모든 엔드포인트의 request 상태에 대한 유효성 검증 추가

feat: 예약 내역 조회 시 시작 날짜, 종료 날짜 순서 검증 로직 추가

refactor: 예약 내역 조회 엔드포인트 URI 변경

- /admin/reservations -> /reservations/filtered

chore: 초기 데이터 수정

feat: JWT에 email이 아닌 id(PK)를 포함시키도록 수정, MemberRepository 커맨드와 쿼리 분리

refactor: 커스텀 예외 패키지 구조 변경, 전체 예외를 잡는 ExceptionHandler 추가

test: 테스트 픽스처 수정

refactor: 회원가입 성공 시 응답하는 상태 코드 수정 (OK -> CREATED)

refactor: 예약 dto 요청, 응답 패키지 분리

test: 테스트 픽스처 패키지 이동

test: 회원 추가 API 테스트 추가

test: 로그인, 로그인 체크 API 테스트 추가

test: 회원가입, 로그인 테스트 픽스처 추가

feat: 권한이 필요한 엔드포인트에 대해 유효하지 않거나 권한 없는 JWT로 요청 시 예외 처리하도록 수정

test: 테마 API 테스트 추가

chore: 테스트명 변경

test: 예약 시간 API 테스트 추가

test: 예약 API 테스트 추가

chore: 테스트명 일관되게 변경

fix(프론트엔드): 예약 생성 요청 엔드포인트 수정

fix: orElseGet을 orElse로 수정

remove: 사용하지 않는 기능 제거

fix(프론트엔드): 잘못된 상태 코드 비교 로직 수정

fix: 상황에 적합한 예외로 수정

feat: HttpServletRequest로부터 accessToken 추출할 때 null 검증 추가

fix: 요청 dto 누락된 NotNull 검증 추가

fix: 잘못된 예외 메시지 수정

fix: 일반 회원인 경우, 본인의 예약 내역만 삭제할 수 있도록 수정

fix: 메서드명 오타 수정

chore: 예외 메시지를 구체적으로 수정

refactor: AuthRole.Member 한글 설명 변경

refactor: PostConstruct 사용 없이 SecretKey 값을 주입하도록 수정

comment: TODO 추가

refactor: 메서드명 개선(delete -> deleteById)

refactor: JWT 파싱 예외 구체적으로 분리 (학습용)

refactor: ExceptionHandler가 적용된 메서드명 일관되게 수정

chore: LocalTime 응답 형식 일관되게 수정

comment: TODO 추가

test: 클라이언트가 인가되지 않은 경우, UNAUTHORIZATION이 아닌 FORBIDDEN를 응답하도록 수정했었음. 프로덕션 코드 수정 후 고치지 않은 테스트 코드 수정.

-

test: 예약 시간, 테마, 회원 추가 행위에 대한 테스트 픽스처 추가

feat: 이메일 형식 검증 강화

- OWASP Validation Regex Repository를 따름

chore: final 키워드 추가

fix: 잘못 설정한 given 수정

test: Member 생성 검증 테스트 추가

chore: 해결된 TODO 제거

refactor: for 대신 stream 사용을 사용하여 가독성 개선

feat(Member): equals, hashCode 재정의

feat: 예약 삭제 기능 검증 수정

- 삭제 권한을 application 계층에서 검증하도록 수정

fix: 관리자는 항상 예약 삭제가 가능하도록 수정

refactor: 불필요한 오토 언방식을 방지하기 위해 반환 타입 통일 (boolean -> Boolean)

refactor: service 계층에서 인기 테마의 기준을 조절할 수 있도록 수정

feat: 요청에 토큰이 있어야 하는데 없는 경우, JwtTokenExtractor에서 null을 반환하는 대신 커스텀 예외를 발생시키도록 수정

chore: members 테이블의 role 컬럼 CHECK 제약 조건 추가

feat(AuthService): 사용자를 찾지 못했을 때 예외를 발생시키도록 수정

chore: 조금 더 명확한 예외로 수정

- ResourceNotFoundException -> MemberNotFoundException

perform: 예약 가능 시간 조회 기능 성능 개선

- 한 번의 쿼리로 특정 날짜와 테마에 대한 모든 예약된 시간을 조회하도록 개선

fix: AuthTokenNotFoundException 예외 처리를 위한 ExceptionHandler 추가

refactor: ThemeRepository 커맨드와 쿼리 분리

feat(MemberAuthInfo): JWT의 payload에 담기는 Member의 정보에 대응되는 클래스 추가 및 적용

chore: Service 예외 메시지 일관되게 수정

remove: 불필요한 커스텀 예외 제거

docs: API 응답 상태 코드 수정, 표 나열 순서 정리(같은 리소스에 대해 POST, DELETE, GET 순서)

feat: 과거 시간으로 예약을 하지 못하도록 검증 추가

docs: 과거 시간 예약 추가에 대한 검증 설명 추가

refactor(ResourceInUseException): 상황에 더 적절한 커스텀 예외를 호출하도록 수정

refactor: Optional.of 사용하여 가독성 개선

chore: 이메일 정규표현식 상수화

refactor: 예약, 예약 시간 Repository 커맨드와 쿼리 분리

test: 커스텀 빈 주입을 위한 TestConfig 파일 분리

chore: 불필요한 주석 제거

feat: repository에 getById 기능 추가 및 적용

refactor: 예약 추가 기능 메서드 분리

feat: 여러 행이 조회되는 경우 예외 처리

refactor(AuthRoleCheckInterceptor): 인증,인가 검증 구조 및 순서 개선

remove: 사용되지 않는 기능 제거

junhaa committed:
fix: 인터셉터 인증 로직 수정

junhaa committed:
chore: JPA 의존성 추가

junhaa committed:
feat: 예약 시간 repository JdbcTemplate -> JPA로 전환

junhaa committed:
feat: 테마 repository JdbcTemplate -> JPA로 전환

junhaa committed:
feat: 회원 repository JdbcTemplate -> JPA로 전환

junhaa committed:
feat: 예약 repository JdbcTemplate -> JPA로 전환

junhaa committed:
chore: 썸네일 URL 수정

junhaa committed:
test: JPA 마이그레이션 후 깨진 테스트 수정

junhaa committed:
test: 테스트 명 언더바를 공백으로 치환하여 Test Tree 가독성 개선

junhaa committed:
feat: 예약 상태 enum 추가

junhaa committed:
feat: 예약 상태 목록 조회 기능 추가

junhaa committed:
feat: 관리자 예약 추가 로직 수정, 내 예약 목록 조회 기능 추가

junhaa committed:
feat(프론트엔드): 관리자 예약 추가 로직 수정, 내 예약 목록 조회 기능 추가

junhaa committed:
test: 내 예약 목록 조회 기능 추가로 인해 깨진 테스트 수정

junhaa committed:
test: 예약 상태 목록 조회 기능 테스트 추가

junhaa committed:
test: 특정 회원의 예약 목록 조회 기능 테스트 추가

threepebbles committed:
chore: 메서드 선언 순서 변경

threepebbles committed:
docs: 상태 목록 조회 API 추가, 변경된 API 스펙 수정

threepebbles committed:
Merge remote-tracking branch 'remotes/woowa/threepebbles' into step1

threepebbles committed:
fix: 전체 예약 목록 조회 테스트 수정

threepebbles committed:
remove: 사용하지 않는 기능 제거

threepebbles committed:
feat(RequiresRole): 클래스에도 어노테이션 추가 가능하도록 수정

threepebbles committed:
refactor(AdminReservationRestController): 관리자 전용 RestController 분리

threepebbles committed:
refactor(프론트엔드): 관리자 예약 목록 조회 엔드포인트를 상수를 사용하여 정의

threepebbles committed:
refactor: 관리자 권한 기반의 엔드포인트와 회원 권한 기반의 엔드포인트를 관리하는 컨트롤러 분리

threepebbles committed:
test: 관리자 권한 기반의 엔드포인트와 회원 권한 기반의 엔드포인트를 관리하는 컨트롤러 테스트 분리

threepebbles commented:
@CodeRabbit Why some checkes were not successful? Could I know the reason?

threepebbles committed:
remove: 사용하지 않는 기능 제거

threepebbles committed:
test: 과거 시간으로 예약 추가 테스트 추가

threepebbles commented:
@CodeRabbit js와 html 파일은 리뷰하지 말아줘

summerlunaa commented on build.gradle:
```diff
@@ -35,6 +35,9 @@ dependencies {
+    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
+
     // h2
     runtimeOnly 'com.h2database:h2'
```
jdbc, h2는 왜 남겨두신건가요?

threepebbles replied:
사용하지 않는 jdbc 의존성은 삭제했습니다!
h2는 프로덕션 코드와 테스트 코드에서 아직 사용 중인 기능이므로 남겨두었습니다.

- 프로덕션 코드의 application.yml
```yml
  datasource:
    url: jdbc:h2:mem:database
    username: sa
    password: password
```

- 테스트 코드의 application.yml
```yml
  datasource:
    url: jdbc:h2:mem:testdb
    username: test
    password: password
```

summerlunaa replied:
h2는 당연히 있어야하는데 잘못 여쭤봤네욥ㅎㅎ 확인 감사합니다~

summerlunaa commented on src/main/java/roomescape/member/application/MemberService.java:
```diff
@@ -18,18 +20,23 @@ public class MemberService {
+
+        final Long id = memberCommandRepository.save(member);
+        final Member found = memberQueryRepository.findById(id)
+                .orElseThrow(() -> new ResourceNotFoundException("해당 회원을 찾을 수 없습니다."));
```
member를 다시 조회하는 이유가 무엇인가요~?

threepebbles replied:
개발 초기에 레포지토리(`xxxCommandRepository`)에서 도메인 객체를 저장하는 `save()` 메서드의 반환 값을 전부 Long id(PK)로 했었습니다. save한 객체의 전체 정보를 필요로 하는 경우보다 save만 하는 경우가 많아질 것이라 예상해서 그렇게 설계했었는데요!

요구사항대로 구현하고 보니 대부분의 응답에서 id 이상의 정보를 필요로 해서 이런 비효율적인 구조(save후 find로 정보를 얻어오는 구조)가 자주 보이게 되었네요 😞 

모든 레포지토리에서 save 후 Long id가 아닌 저장한 도메인 객체를 반환하도록 수정했습니다.

반영 커밋: 93c187aae3d6052d7b0e5e0d1e6c23fbde01e521, a43973572a9e4bcb745bdf81597d1a83f718824c, f8d99a5045870ab31ff0a2cf6202905b0367d86d

summerlunaa commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -1,22 +1,42 @@
 @EqualsAndHashCode(of = {"id"})
 public class Member {
 
     private static final String EMAIL_REGEX = "^[a-zA-Z0-9_+&*-]+(?:\\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,}$";
```
정규식을 사용할 때 어떤 점에 주의해야 할까요? [블로그](https://blog2.deliwind.com/20200430/Java-%EC%A0%95%EA%B7%9C%ED%91%9C%ED%98%84%EC%8B%9D%EC%9D%84-%EC%82%AC%EC%9A%A9%ED%95%A0-%EB%95%8C-%EC%A3%BC%EC%9D%98%ED%95%A0-%EC%A0%90/)

threepebbles replied:
```java
// String.class
public final class String {
    ...
    public boolean matches(String regex) {
        return Pattern.matches(regex, this);
    }
    ...
}
// Pattern.class
public final class Pattern {
    ...
    public static boolean matches(String regex, CharSequence input) {
        Pattern p = Pattern.compile(regex);
        Matcher m = p.matcher(input);
        return m.matches();
    }
    ...
}
```
전혀 모르고 있었는데 `String.matches()`를 호출하면 내부적으로 매번 Pattern 객체를 생성해서 패턴 매칭을 하고 있었군요!
Pattern 객체를 캐싱해서 사용하도록 수정했습니다. 감사합니다 😄 

반영 커밋: e3fde9f817efd34770c4789afd00c21bfcc59d26

summerlunaa commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -1,22 +1,42 @@
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    @Column(nullable = false, unique = true)
```
Column 애노테이션 잘 활용하셨네요 👍

summerlunaa commented on src/main/java/roomescape/member/infrastructure/MemberRepositoryImpl.java:
```diff
@@ -0,0 +1,41 @@
+
+@Repository
+@RequiredArgsConstructor
+public class MemberRepositoryImpl implements MemberCommandRepository, MemberQueryRepository {
```
JpaRepository만 사용해도 될 것 같은데 CommandRepo, QueryRepo 인터페이스와 구현체는 왜 필요한가요? 이러한 설계를 선택한 이유와 장단점이 궁금합니다.

threepebbles replied:
이전 미션에서 최상위 패키지를 핵심 도메인 별로 나누고(ex. member, reservation, theme, auth) 같은 패키지 내에서는 응집도를 높이면서 다른 패키지와의 결합도는 낮추는 설계 방식을 고민해보았었습니다.

그 고민의 결과 중 하나가 다른 도메인에서 엔티티의 조회(Query)만 허용하고 커맨드(Command)는 허용하지 않게 하는 것입니다. 예를 들면 DB에 저장되는 Theme 엔티티의 상태는 반드시 theme 패키지 하위에서만 수정/변경이 일어나야 하고, 다른 도메인 패키지 하위의 기능에 의해 상태가 변경되면 안 된다는 기준을 만들었습니다. 계층 간 의존 구조를 A도메인의 Service에서 B도메인의 QueryRepository에만 의존하게 했습니다.

사이드 이펙트를 발생시키는 행위를 해당 패키지 내부에서만 가능하게 함으로써, `엔티티의 상태 관리에 대한 책임`을 최상위 패키지 단위로 명확하게 분리하고자 했습니다.

페어 프로그래밍 당시에는 JDBC->JPA 마이그레이션을 하게 되면서 dirty checking이 발생한다는 것을 생각하지 못했었는데요...! dirty checking으로 인해 조회 해온 엔티티의 상태가 변했을 때 자동으로 flush()가 일어나지 않도록 `@Transactional(readOnly = true)`을 추가했습니다. (반영 커밋: 96b2555020a2d62f5040795ac0ab9e8262c6cdf0)

커맨드와 쿼리를 나눔으로써 패키지 간 결합도를 낮추려고 한 것이 적절한 방향의 고민인지, 적절하다면 현재 코드에서 분리하고자 했던 목적이 충분히 달성되고 있는지(ex. 커맨드 쿼리 인터페이스를 분리하고 `@Transcational(readOnly=true)`를 추가한 것만으로 충분한지) 여쭤보고 싶습니다..! 🙏

summerlunaa replied:
자세한 설명 감사합니다 헤일러! 쿼리와 커맨드를 나눠 다른 도메인에서는 조회만 가능하도록 하려는 의도가 있었군요! 좋습니다. 그런 의도였다면 괜찮은 방식이라 생각합니다. 다만 상위 클래스(ex 서비스)에서 트랜잭션을 `readOnly = false`로 열면 의미가 없을 수도 있겠네요. 더티 체킹 자체를 막으려면 Entity가 아니라 DTO를 반환하는 방법도 있을 것 같은데 이러면 또 Entity가 필요한 곳에서는 사용이 어렵구요. 
Query과 Command를 Repo가 아니라 서비스 레이어에서 나누는 것도 방법일 것 같아요. QueryService와 CommandService로요. 그리고 다른 도메인에서는 QueryService를 사용하고, 본인의 도메인에서는 Repository를 직접 사용하는거죠. 그리고 QueryService가 단순히 Repository의 메서드를 프록싱만 하는 것이 아니라 예외 검증 등의 로직까지 들고 있으면 다른 여러 사용처에서 조금 더 편리하게 Query 메서드들을 사용할 수 있고, QueryService의 역할도 부여할 수 있을 것 같아요. 필요하면 Service에서는 Service용 DTO를 반환할 수도 있구요.
다만 이 모든 것이 '당장' 필요한지는 의문이긴합니다. 이렇게 설계가 복잡해지면 단점도 생기니까요. Repository만 정의해두고 사용하면 간편한데 레이어가 추가되는 바람에 관리해야 할 클래스도 늘어나고 사용하는 메서드가 추가될 때마다 여러 클래스를 변경해주어야 하죠. 레이어도 복잡해져서 새로운 사람이 와서 헤일러의 코드를 볼 때 이해하는 시간이 훨씬 더 걸립니다. 서비스용 DTO를 추가한다면 복잡성과 관리에 드는 비용은 더 커지구요. 이런 모든 단점을 안고도 위 설계를 적용하는 것이 나을지는 헤일러가 고민해보면 좋을 것 같습니다 :)

threepebbles replied:
답변 감사합니다 파랑 :)
답변 주신 내용을 바탕으로 주말 동안 Command, Query 분리 구조를 유지한다면 어떻게 더 적절하게 분리할 수 있을지, 지금 도입할 필요가 있을지 고민해보았습니다.

우선 Command, Query를 유지한다고 했을 때, 클라이언트의 유스케이스 관점에서 Repository 계층에서 나누기 보다는 Service 계층에서 분리하는 것이 더 적절해보였습니다. 말씀해주신 예외 검증의 책임도, Transaction을 관리하는 책임도 Service 계층에서 관리할 때 더 효과적으로 관리될 수 있기 때문입니다.

이렇게 되면 다른 도메인의 Service가 QueryService에 의존하는 구조가 될 것인데, 이 구조에서 Service 간의 의존 방향을 확실하게 구분하기 위해 Service도 두 계층으로 나누어 관리해야 할 필요성을 느꼈습니다. 

DDD에서는 Application Service와  Domain Service로 나누어 Application Service가 사용자 요청 단위의 유스케이스 로직을 담당하고, Domain Service가 엔티티에 속하지 않는 도메인 로직을 담당하도록 역할을 나눈다고 들었습니다. (제가 DDD를 깊이 이해하고 있지는 않고 필요할 때 일부 용어만 차용해서 학습하고 있습니다.)

하지만 현재 요구사항에서는 회원, 예약, 테마에 대한 단순한 CRUD 기능만 제공해주면 되는데, Service 계층을 2개로 나누고 또 CommandService, QueryService까지 분리함으로써 학습적으로 어떤 의미가 있을까 고민이 들었습니다.

지금은 조회 성능을 최적화해 보고 이를 테스트 해볼 수 있는 상황도 아니기 때문에, 학습적으로 얻어가는 것보다 단순 반복 작업으로 인해 유지보수에 드는 비용이 더 커질 것이란 생각이 들었습니다.

그래서 지금은 CQRS를 잠시 접어두고 우선 Service, Repository를 각각 하나의 구현체로 가져가는 방향으로 진행해보려고 합니다 :)

summerlunaa replied:
많은 고민이 있었을 것 같네욥 ㅎㅎ 네 좋습니다 👍 지금은 어플리케이션의 크기가 작은 만큼 설계를 단순하게 가져가보시죠~ 나중에 정말로 서비스 계층의 분리가 필요해지는 순간이 올 거예요! 그 때 수정해도 늦지 않을 것 같습니다.

summerlunaa commented on src/main/java/roomescape/reservation/application/ReservationService.java:
```diff
@@ -34,48 +37,85 @@ public class ReservationService {
-    public ReservationResponse create(final CreateReservationRequest request, final MemberAuthInfo memberAuthInfo) {
-        final ReservationTime reservationTime = getReservationTime(request);
-        validateNoDuplicateReservation(request);
+    public ReservationResponse create(final CreateReservationRequest request) {
```
두 개의 create 메서드의 로직이 거의 동일한 것 같은데 어떻게 생각하시나요~? delete도 마찬가지입니다.

threepebbles replied:
create은 메서드 분리를 통해 중복을 줄였습니다.
반영 커밋: bf8535f7c4e6dfac4e34f671ed45d60ba3c08b9e

delete는 관리자 권한 삭제의 경우 회원 Repository를 조회하지 않고 토큰의 권한만 확인하여 삭제하도록 개선해보았습니다. 
반영 커밋: 50f93209727b9def5f40e887dfc65d581b206ddd

delete는 수정하고 나니 중복이 없어져서 우선 그대로 두었습니다.
혹시 다른 방향으로 개선할 수 있는 점이 있을까요? 🧐

summerlunaa replied:
앞으로 관리자용 메서드가 더 추가될 수도 있는데 모든 메서드에 일괄적으로 ADMIN 검증을 중복으로 넣어줘야 할지는 살짝 고민되는 부분이네요 ㅎㅎ 어드민용 서비스를 분리하는 것도 방법일 것 같구요.

threepebbles replied:
저도 관리자용 예약 추가, 예약 조회 기능과 회원용 예약 추가, 예약 조회 기능이 점점 공통의 로직이 사라지고 각자의 책임을 가지는 것이 자연스러워지고 있다 생각하고 있던 찰나였습니다 ㅎㅎ 그래서 어드민용 서비스도 분리했습니다 😄 

반영 커밋: [feat: 관리자 예약 서비스와 회원 예약 서비스 분리, 관리자는 과거 시간에 예약이 가능하도록 기능 수정](fe795ded7b2a3ae12397732db7c32cefb7003fd8)

summerlunaa commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -1,39 +1,68 @@
+    private Long id;
+    @Column(nullable = false)
+    private LocalDate date;
+    @ManyToOne
+    @JoinColumn(name = "time_id")
+    private ReservationTime time;
+    @ManyToOne
+    @JoinColumn(name = "theme_id")
+    private Theme theme;
+    @ManyToOne
+    @JoinColumn(name = "member_id")
+    private Member member;
```
time, theme, member 모두 ManyToOne으로 설정하셨네요! 다른 옵션은 고려해보지 않으셨는지, 왜 ManyToOne을 선택하게 되었는지 궁금해요!

threepebbles replied:
Reservation과 ReservationTime, Theme, Member의 관계를 생각하면, Reservation에서 참조 하는 경우가 많고, 반대 방향으로 참조하는 경우는 거의 없었습니다. 그래서 Reservation에 대응되는 테이블이 연관관계의 주인으로써 외래키를 보유하도록 하는 것이 적절하다고 생각했습니다! 

다른 더 적절한 옵션이 있을지는 잘 모르겠습니다. 😿

summerlunaa replied:
설명 감사합니다! 더 적절한 옵션이 있어 드린 코멘트는 아닙니다. 잘 선택하셨습니다 👍 
cascade는 어떤 설정인가요? default는 어떻게 동작하나요?

threepebbles replied:
![Untitled](https://github.com/user-attachments/assets/c095da6e-f9f5-4c10-8322-2a25a1fc594b)

JPA에서의 cascade는 영속성 전이의 범위를 정하는 설정입니다. 특정 엔티티를 영속 상태로 만들거나 제거할 때 참조 관계로 연관된 엔티티도 함께 처리 되도록 만드는 것을 `영속성 전이`라고 부릅니다. 

영속성 컨텍스트 내에서 부모 엔티티의 생명주기와 자식 엔티티의 생명주기를 함께 관리하고 싶을 때 사용합니다.

JPA에서 제공하는 cascade 종류는 다음과 같습니다.

| CascadeType | 설명                                  |
| --------- | ----------------------------------- |
| `PERSIST` | 부모가 `persist()` 될 때 자식도 `persist()`됨 |
| `MERGE`   | 부모가 `merge()` 될 때 자식도 `merge()`됨     |
| `REMOVE`  | 부모가 `remove()` 될 때 자식도 `remove()`됨   |
| `REFRESH` | 부모가 DB로부터 다시 로딩될 때, 자식도 다시 로딩됨 |
| `DETACH`  | 부모가 `detach()` 될 때 자식도 `detach()`됨   |
| `ALL`     | 위의 모든 옵션을 포함. 부모와 자식은 모든 라이프 사이클을 함께함                         |

보통 OneToMany 관계에서 `One`을 부모, `Many`를 자식으로 두어 부모가 cascade 옵션을 갖고 자식을 관리하는 구조를 만들 때 사용합니다. 예를 들어 게시글을 저장할 때 첨부 파일도 함께 저장하고, 게시글을 삭제할 때  첨부된 파일도 함께 삭제하고 싶을 때 다음과 같이 설정할 수 있습니다.

```java
// 부모
@Entity
public class Post {
	...
    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL)
    private List<Attachment> attachments;
}

// 자식
@Entity
public class Attachment {
	...
    @ManyToOne
    private Post post;
}
```

자식 엔티티가 여러 부모에게 속해 있거나, 부모 엔티티 없이도 독립적으로 재사용되는 경우에는 cascade 설정을 피해야 하므로 설정에 신중해야 합니다. 그리고 DB의 `ON DELETE RESTRICT`, `ON DELETE CASCADE`와 같은 제약 조건과 충돌할 수 있기 때문에 JPA cacade 설정과 DB 제약 조건 중 하나를 선택하고 일관된 방식으로 사용해야 합니다.

threepebbles replied:
> cascade는 어떤 설정인가요?

![Untitled](https://github.com/user-attachments/assets/c095da6e-f9f5-4c10-8322-2a25a1fc594b)

JPA에서의 cascade는 영속성 전이의 범위를 정하는 설정입니다. 특정 엔티티를 영속 상태로 만들거나 제거할 때 참조 관계로 연관된 엔티티도 함께 처리 되도록 만드는 것을 `영속성 전이`라고 부릅니다. 

영속성 컨텍스트 내에서 부모 엔티티의 생명주기와 자식 엔티티의 생명주기를 함께 관리하고 싶을 때 사용합니다.

JPA에서 제공하는 cascade 종류는 다음과 같습니다.

| CascadeType | 설명                                  |
| --------- | ----------------------------------- |
| `PERSIST` | 부모가 `persist()` 될 때 자식도 `persist()`됨 |
| `MERGE`   | 부모가 `merge()` 될 때 자식도 `merge()`됨     |
| `REMOVE`  | 부모가 `remove()` 될 때 자식도 `remove()`됨   |
| `REFRESH` | 부모가 DB로부터 다시 로딩될 때, 자식도 다시 로딩됨 |
| `DETACH`  | 부모가 `detach()` 될 때 자식도 `detach()`됨   |
| `ALL`     | 위의 모든 옵션을 포함. 부모와 자식은 모든 생명주기를 함께함                         |

보통 OneToMany 관계에서 `One`을 부모, `Many`를 자식으로 두어 부모가 cascade 옵션을 갖고 자식을 관리하는 구조를 만들 때 사용합니다. 예를 들어 게시글을 저장할 때 첨부 파일도 함께 저장하고, 게시글을 삭제할 때  첨부된 파일도 함께 삭제하고 싶을 때 다음과 같이 설정할 수 있습니다.

```java
// 부모
@Entity
public class Post {
	...
    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL)
    private List<Attachment> attachments;
}

// 자식
@Entity
public class Attachment {
	...
    @ManyToOne
    private Post post;
}

```

---
> default는 어떻게 동작하나요?

cascade 설정의 기본값(default)은 아무 영속성 전이가 일어나지 않는 상태(` CascadeType[] cascade() default {};`)입니다.

그리고 cascade 설정을 할 때는 아래와 같은 상황을 고려해서 항상 신중하게 사용해야 합니다.

- 자식 엔티티가 여러 부모에게 속할 수 있는 구조거나, 부모 없이도 독립적으로 재사용 되어야 하는 경우에는 cascade 설정을 피해야 합니다.
- JPA의 cacade 설정과 DB의 `ON DELETE RESTRICT`, `ON DELETE CASCADE`와 같은 제약 조건이 충돌할 수 있기 때문에, 둘 중 하나의 방법을 선택해서 일관되게 사용해야 합니다.

summerlunaa commented on src/main/java/roomescape/reservation/domain/ReservationStatus.java:
```diff
@@ -0,0 +1,16 @@
+import lombok.Getter;
+
+@Getter
+public enum ReservationStatus {
+
+    CONFIRMED("예약"),
+    WAITING("대기");
```
Enum 사용 좋습니다 👍

summerlunaa commented on src/main/java/roomescape/reservation/infrastructure/JpaReservationRepository.java:
```diff
@@ -0,0 +1,33 @@
+    @Query("""
+                SELECT r
+                FROM Reservation r
+                JOIN FETCH r.time t
+                JOIN FETCH r.theme th
+                JOIN FETCH r.member m
```
`JOIN FETCH`는 일반적인 `JOIN`과 무엇이 다른가요?

threepebbles replied:
Reservation 엔티티가 Theme 엔티티를 필드로 갖고 있는 관계에서, DB를 통해 Reservation 엔티티를 조회해올 때 Theme 엔티티의 상태도 즉시 로딩을 할 지, 프록시 객체를 이용해 지연 로딩을 할 지 전략을 결정해야 합니다.  `JOIN FECTH r.theme th`를 사용하면 조회할 때 LEFT JOIN을 이용하여 Theme 객체를 즉시 로딩해옵니다.

현재 ManyToOne설정에서는 FetchType.EAGER가 디폴트기 때문에 `JOIN FETCH`가 아닌 `JOIN`을 해도 즉시 로딩이 될 것으로 기대되지만, JPQL에서 명시적으로 `JOIN FETCH`을 사용하면 추후 `@ManyToOne(fetch = FetchType.LAZY) Theme theme`으로 코드가 변경되더라도 해당 JPQL을 실행하는 메서드에서는 LEFT JOIN + 즉시 로딩하도록 강제할 수 있습니다. 😄

summerlunaa commented on src/main/java/roomescape/reservation/ui/ReservationStatusRestController.java:
```diff
@@ -0,0 +1,27 @@
+
+    private final ReservationStatusService reservationStatusService;
+
+    @GetMapping("/statuses")
```
path를 `/statuses`로 정하면 어떤 상태를 조회하는 API인지 path만 보고도 알 수 있을까요?
StatusService도 확인해보니 굳이 필요한가? 하는 생각이 들더라구요. status를 reservation과 따로 분리할 필요가 있는지 한번 고민해보면 좋을 것 같습니다 :)

threepebbles replied:
> path를 /statuses로 정하면 어떤 상태를 조회하는 API인지 path만 보고도 알 수 있을까요?

제 생각에도 `/statuses`만으로는 예약 상태 목록을 조회한다는 의미가 잘 드러나지 않는다는 생각이 들어 조금 더 명확하게 `/reservation-statuses`로 URI를 변경했습니다!

반영 커밋: 37ce82fc3b7cb7d4592f77c8530b74f87a6fb03c


> StatusService도 확인해보니 굳이 필요한가? 하는 생각이 들더라구요. status를 reservation과 따로 분리할 필요가 있는지 한번 고민해보면 좋을 것 같습니다 :)

해당 기능은 관리자 예약 관리 페이지에서 상태도 선택하여 예약을 추가할 수 있도록 하기 위해 필요한 기능이어서 추가했었는데요! 프론트에서 필요로 하는 요구 사항에 따라 빠르게 구현하다 보니 크게 고민 없이 StatusService를 만든 것 같습니다. 😓 제 생각에도 reservation과 status를 따로 분리할 필요가 없어 보여서 하나로 합쳤습니다.

![image](https://github.com/user-attachments/assets/7445a859-c971-41c0-87f5-6592f544dc91)

반영 커밋: 80a98531fa2412360ac81cebe97f4b491fa58746

summerlunaa commented on src/main/java/roomescape/reservation/ui/dto/response/ReservationResponse.java:
```diff
@@ -19,6 +21,27 @@ public static ReservationResponse from(final Reservation reservation) {
+        );
+    }
+
+    public record ForMember(
+            Long id,
+            String theme,
+            LocalDate date,
+            LocalTime time,
+            String status
+    ) {
```
Member용 응답 클래스를 따로 만든 이유가 있을까요?

threepebbles replied:
우선 관리자의 조회 API에서 필요로 하는 응답과 일반 회원의 조회 API에서 필요로 하는 응답이 다르기 때문에 분리했습니다. 권한에 따라 조회할 수 있는 정보의 범위가 다를 수 있기 때문에, 응답은 항상 필요한 만큼의 정보만 담고 있어야 한다고 생각합니다.

조금 더 나아가서 response dto를 내부 클래스로 설계를 하게 된데는 나름의 이유가 있는데요!
제가 했던 고민을 조금 더 적절하게 설명할 수 있는 예시를 가져왔습니다.
```java
public record MemberResponse(
        Long id,
        String name,
        String email

) {

    public record IdName(
            Long id,
            String name
    ) {
        public static IdName from(final Member member) {
            return new IdName(member.getId(), member.getName());
        }
    }

    public record Name(
            String name
    ) {
        public static Name from(final Member member) {
            return new Name(member.getName());
        }
    }
}
```

request-response dto 쌍을 설계하면서 느낀 점이 있었는데요!
request dto는 특정 엔드포인트에서만 사용되기 때문에 재사용되는 경우가 거의 없었습니다. 반면, **response dto는** 도메인 객체의 일부 상태만 응답하는 요구가 생기면서 **재사용할 수 있는 여지가 많았던 것 같습니다.**

예를 들어 `POST /members`, `GET /members` 엔드포인트에서는 Member의 id, name만 응답하면 되고, `GET /login/check` 엔드포인트에서는 name만 응답하면 됩니다.

이때 부분 상태의 조합마다 매번 별도의 dto 클래스를 만들게 되면 최악에 $2^{상태 개수}$가지 응답 dto의 조합이 나올 수 있습니다. 이렇게 되면 파일 개수가 지나치게 많아져 가독성과 유지보수성이 떨어진다는 문제점이 생긴다고 생각했습니다.

그래서 Member의 전체 상태 정보를 가진 MemberResponse를 기준으로 삼고, 부분 상태를 응답하기 위한 dto를 내부 클래스 형태로 만들어서 사용했습니다. 

이렇게 하면 Member의 응답 dto를 하나의 파일로 묶어낼 수 있어 응집도가 높아지고 관리도 쉬워진다 생각했습니다 :)

summerlunaa commented on src/main/java/roomescape/theme/infrastructure/JpaThemeRepository.java:
```diff
@@ -0,0 +1,33 @@
+                    t.name AS name,
+                    t.description AS description,
+                    t.thumbnail AS thumbnail,
+                    COUNT(r.id) AS reservation_count
```
count는 필요한 값인가요?

threepebbles replied:
```java
    @Query(value = """
                SELECT
                    t.id AS id,
                    t.name AS name,
                    t.description AS description,
                    t.thumbnail AS thumbnail,
                    COUNT(r.id) AS reservation_count
                FROM themes t
                LEFT JOIN reservations r 
                    ON t.id = r.theme_id AND r.date BETWEEN ? AND ?
                GROUP BY t.id, t.name, t.description, t.thumbnail
                ORDER BY reservation_count DESC
                LIMIT ?;
                """, nativeQuery = true)
    List<Theme> findTopNThemesByReservationCountInDateRange(
            final LocalDate dateFrom,
            final LocalDate dateTo,
            final int limit
    );
```
`COUNT(r.id) AS reservation_count`는 해당 테마로 예약된 예약의 개수를 의미하는 컬럼입니다. `ORDER BY`를 통해 예약 개수 순으로 내림차순 정렬할 기준을 주기 위해 사용했습니다. 😀 

혹시 count를 조회 테이블의 컬럼으로 만들지 않고, `ORDER BY`의 정렬 기준으로 줄 수 있는 방법이 있을까요?

threepebbles replied:
그리고 방금 생각이 났는데요! 
`findTopNThemesByReservationCountInDateRange()`의 반환 타입이 Theme인데, Theme이 count를 필드로 갖고 있지 않아 매핑 과정에서 발생할 수 있는 문제가 있을까요?

summerlunaa replied:
아하 넵넵 이해했습니다. 결과(Theme)에서 count 값을 사용하지 않아서 드린 질문인데 order by에서 사용중이었군요.

summerlunaa commented on src/main/resources/application.yml:
```diff
@@ -8,12 +8,14 @@ spring:
-      mode: embedded
-      schema-locations: classpath:schema.sql
-      data-locations: classpath:data.sql
+  jpa:
+    hibernate:
+      ddl-auto: create-drop
+    defer-datasource-initialization: true
+    show-sql: true
+    properties:
+      hibernate:
+        format_sql: true
```
각각의 설정들은 어떤 의미/역할인가요?

threepebbles replied:
각 설정은 다음과 같은 이유로 사용했습니다!

```yml
  jpa:
    hibernate:
      ddl-auto: create-drop # 애플리케이션 실행 시작 시기에 @Entity 정보를 기반으로 테이블을 생성하고, 종료 시 삭제함
    defer-datasource-initialization: true # JPA가 @Entity 정보를 기반으로 DDL을 실행한 이후에 data.sql이 실행됨을 보장함
    show-sql: true # JPA가 실행하는 SQL 쿼리를 콘솔에 출력해주는 기능. 디버그 용도.
    properties:
      hibernate:
        format_sql: true # 출력되는 SQL 쿼리를 읽기 편하게 포맷팅 해주는 기능(들여쓰기, 공백 이쁘게 해주는 역할)
```

summerlunaa replied:
실제 운영하는 애플리케이션이라면 위 설정을 어떻게 수정하는 게 좋을까요? :)

threepebbles replied:
웹 애플리케이션을 실제로 운영해본 경험이 없다보니, 검색과 함께 상상의 나래를 펼쳐보았습니다 ㅎㅎ..

### `ddl-auto: none`
애플리케이션 재시작만으로 고객들의 소중한 데이터가 일부 혹은 전부 손실 될 수 있으므로 테이블 스키마는 처음 애플리케이션 시작 시 설정한 그대로 계속 유지해야 할 것 같습니다.

### `defer-datasource-initialization: false`
`data.sql`은 테스트 환경, 개발 환경에서 사용하는 설정이고 운영 환경에서 사용하는 경우는 거의 없을 것 같습니다.

### `show-sql: false`
- SQL 로그에는 고객들의 민감 정보가 담겨있을 수 있어 모든 쿼리가 출력 되는 것 자체가 보안상 위험합니다.
- 트래픽이 많은 운영 환경에서는 개발자가 실시간으로 콘솔 출력을 보며 디버깅 할 수 없습니다. 중요한 쿼리만 필터링해 로그 파일로 남기는 것이 더 효과적입니다.
- 콘솔에 문자열을 출력하는 비용이 매우 큽니다. (리소스 낭비)

위와 같은 이유로 운영 환경에서는 콘솔에 SQL 로그를 찍어볼 이유는 없을 것 같습니다.

### `format_sql: false`
콘솔 출력을 예쁘게 해줄 뿐인 설정이기 때문에 `show-sql`와 함께 키고 꺼야 하는 옵션입니다.

summerlunaa replied:
ㅎㅎ 정리를 너무 잘해주셨네요 보통은 말씀하신 그대로 설정합니다 👍

summerlunaa commented on src/test/java/roomescape/fixture/config/TestConfig.java:
```diff
@@ -0,0 +1,123 @@
+import roomescape.reservation.infrastructure.JpaReservationTimeRepository;
+import roomescape.reservation.infrastructure.ReservationRepositoryImpl;
+import roomescape.reservation.infrastructure.ReservationTimeRepositoryImpl;
+import roomescape.theme.applcation.ThemeService;
```
패키지명에 오타가 있네요

summerlunaa commented on src/test/java/roomescape/reservation/ui/AdminReservationRestControllerTest.java:
```diff
@@ -0,0 +1,218 @@
+    @Test
+    void 예약을_추가한다() {
+        final Map<String, String> adminCookies = adminLoginAndGetCookies();
+        final Map<String, String> reservationParams = confirmedReservationParams();
```
요청 객체로 Map이 아닌 RequesetDto를 사용하면 더 편리하게 테스트 요청을 만들 수 있습니다!

threepebbles replied:
RestAssured로 HTTP 요청 메시지를 조립할 때, request dto 타입을 그대로 사용할 수 있다는 사실을 이번에 알게 되었네요!
타입 안정성도 높아지고 유지보수 하기도 쉬워진 것 같아요. 감사합니다! 😀

반영 커밋: 5e151ca2732ff2ab2bce744d2847de7ca4f3bae3

summerlunaa commented on src/main/java/roomescape/reservation/infrastructure/JpaReservationRepository.java:
```diff
@@ -0,0 +1,33 @@
+import org.springframework.data.jpa.repository.Query;
+import roomescape.reservation.domain.Reservation;
+
+public interface JpaReservationRepository extends JpaRepository<Reservation, Long> {
```
Repository에 대한 테스트는 불필요한가요? 쿼리가 의도한대로 동작하는지 확신할 수 있나요?

threepebbles replied:
제가 생각하기에 Repository에 대한 테스트는 곧 SQL 쿼리문에 대한 테스트라고 생각합니다.

JpaRepository가 기본으로 제공해주는 save(), findById()와 같은 메서드에 대해서는 Spring Data JPA의 구현을 신뢰하고 사용해도 괜찮다고 생각합니다. 

다만 이름이 길고 로직이 복잡한 쿼리 메서드, `@Query`를 통해 직접 정의한 JPQL이나 Native Query는 쿼리가 의도한 대로 동작하는지 테스트가 필요할 수 있을 것 같습니다.

현재 요구사항에서는 `인기테마 목록 반환` 기능이 제가 생각하기에 테스트가 필요한 기준에 부합하는 것 같은데요,

파랑은 어떤 관점으로 Repository의 어떤 부분에 대해 테스트가 필요하다 생각하시는지 궁금합니다!


반영 커밋: [인기 테마 조회 기능 테스트 추가](8b888c753af40196bb17469d963a500973e62b30)

summerlunaa replied:
> 다만 이름이 길고 로직이 복잡한 쿼리 메서드, @Query를 통해 직접 정의한 JPQL이나 Native Query는 쿼리가 의도한 대로 동작하는지 테스트가 필요할 수 있을 것 같습니다.

저도 동의합니다. 기본 메서드의 경우 어느정도 믿고 사용할 수 있지만 헤일러가 말씀해주신 케이스라면 테스트를 작성하는 것이 더 좋겠네요~

summerlunaa reviewed (changes requested):
헤일러 안녕하세요! 파랑입니다 🐳 만나서 반가워요~~
이미 JPA에 익숙하시군요ㅎㅎ 작성하신 코드와 남겨주신 질문에 대해 코멘트 드렸습니다.
혹시 어떤 주제든 질문이 있거나 리뷰에서 이해 안 되는 부분 있으면 언제든 편하게 DM 주셔도 괜찮습니다~ 이번 미션도 파이팅 해보시죠 💪

threepebbles committed:
fix: camelCase로 작성되지 않은 클래스명 오타 수정

threepebbles committed:
chore: 사용하지 않는 jdbc 의존성 제거

threepebbles committed:
refactor: boolean 타입을 반환함을 더 명확하게 예측할 수 있는 메서드명으로 변경

threepebbles committed:
refactor(MemberCommandRepository): save() 메서드 반환 값 수정

- Long id -> Member savedMember

threepebbles committed:
test: MemberService 테스트 추가

- 회원 추가, 회원 삭제 테스트 추가

threepebbles committed:
refactor(MemberCommandRepository): save() 메서드 반환 값 수정

- Long id -> Theme savedTheme

threepebbles committed:
refactor(ReservationCommandRepository, ReservationTimeCommandRepository): save() 메서드 반환 값 수정

- Long id -> Reservation savedReservation
- Long id -> ReservationTime savedReservationTime

threepebbles committed:
perform: 정규표현식을 Pattern 타입으로 캐싱해서 성능 개선

threepebbles committed:
feat: 조회 기능 메서드에 @Transactional(readOnly = true) 추가

threepebbles committed:
chore: 공백 추가로 가독성 개선

threepebbles committed:
refactor: 예약 상태 목록 조회 엔드포인트 URI 경로 개선

- /statuses -> /reservation-statuses

threepebbles committed:
refactor: ReservationStatusService 기능 ReservationService로 통합

threepebbles committed:
chore(application.yml): 설정의 역할에 대한 설명을 주석으로 추가

threepebbles committed:
refactor: dto 이름 변경

- CreateMemberRequest -> SignUpRequest

threepebbles committed:
refactor: dto 이름 변경

- CreateAccessTokenRequest -> LoginRequest

threepebbles committed:
test: HTTP 요청 body 테스트 픽스처 타입 변경

 - String -> request dto 타입으로 변경

threepebbles committed:
chore: 불필요한 개행 제거

threepebbles committed:
test: 예약 시간, 테마 테스트 픽스처 개선

- 개수를 인자로 받아 개수만큼 생성하는 메서드 추가

threepebbles committed:
test(ThemeRepositoryImpl): 인기 테마 조회 기능 테스트 추가

threepebbles committed:
test: Test Tree 가독성 개선을 위해 @DisplayNameGeneration(ReplaceUnderscores.class) 추가

threepebbles committed:
chore: 예외 메시지 수정

threepebbles committed:
refactor: createReservation() 메서드 분리

threepebbles committed:
docs: 예약 상태 목록 조회 URI 수정

threepebbles committed:
refactor: 관리자 권한 예약 삭제 기능 로직 수정

threepebbles committed:
refactor: 관리자 권한 예약 삭제 기능 메서드명 수정

threepebbles committed:
chore: AuthRole 클래스 static import

threepebbles commented:
안녕하세요 파랑! 리뷰 해주신 내용들이 많은 도움이 되었습니다. 
코드가 전보다 깔끔해지고 있는 게 느껴지네요. 😀
이번에도 리뷰 잘 부탁드립니다!

threepebbles committed:
chore: 테스트 메서드명 개선

threepebbles commented on src/main/java/roomescape/auth/config/WebMvcConfig.java:
```diff
@@ -0,0 +1,33 @@
+    private final AuthTokenProvider authTokenProvider;
+
+    @Override
+    public void addInterceptors(final InterceptorRegistry registry) {
+        registry.addInterceptor(
+                        new AuthRoleCheckInterceptor(authTokenExtractor, authTokenProvider)
+                )
+                .addPathPatterns("/**");
+    }
```
곧 해결하러 갈게 좀만 기다려

threepebbles committed:
refactor: AuthRoleCheckInterceptor 적용 범위 수정

- 정적 리소스와 공개 페이지에 대해서는 Interceptor를 거치지 않도록 수정

threepebbles committed:
fix: 패키지명 오타 수정

threepebbles commented:
리뷰 요청 후에 깜빡하고 반영하지 않은 피드백이 생각나서 3개의 커밋을 추가로 했습니다..!
혹시 리뷰 중이셨다면 죄송합니다. 🙇 

아래는 첫 리뷰 요청 이후 변경 사항에 대한 커밋입니다.
[1차 리뷰 요청 이후 변경 사항](https://github.com/woowacourse/spring-roomescape-waiting/pull/177/files/29f862f8549bf22dd8fc85896043fb9ace03c455..78eff7eb06edcb6991253caf551ab15e42c26ab6)

summerlunaa commented on src/test/java/roomescape/theme/infrastructure/ThemeRepositoryImplTest.java:
```diff
@@ -0,0 +1,115 @@
+@DataJpaTest
+@Import(TestConfig.class)
+@DisplayNameGeneration(ReplaceUnderscores.class)
+class ThemeRepositoryImplTest {
```
Dao 테스트 추가 확인했습니다 👍 findAllByThemeIdAndMemberIdAndDateRange 요 메서드도 테스트 추가해주면 좋을 것 같네요~

threepebbles replied:
findAllByThemeIdAndMemberIdAndDateRange도 테스트 추가했습니다!

반영 커밋: 674a3d47475379c9468ebe54af3c4f4708e7f89c

summerlunaa commented on src/main/java/roomescape/reservation/ui/ReservationRestController.java:
```diff
@@ -70,4 +71,11 @@ public ResponseEntity<List<AvailableReservationTimeResponse>> findAvailableReser
         return ResponseEntity.ok(availableReservationTimes);
     }
+
+    @GetMapping("/reservation-statuses")
```
`reservations`로 path prefix를 통일하는 건 어떨까요?

threepebbles replied:
관리자 예약 관리 API와 회원 예약 관리 API를 분리하면서 prefix도 전부 통일해보았습니다!

| HTTP Method | Endpoint                      | AuthRole      | Success          | Description                                                                |
|-------------|-------------------------------|---------------|------------------|----------------------------------------------------------------------------|
| POST        | /reservations                 | ADMIN, MEMBER | 201 CREATED      | 일반 회원의 예약 추가                                                               |
| DELETE      | /reservations/{id}            | ADMIN, MEMBER | 204 NO_CONTENT   | id(PK)에 해당하는 예약 삭제                                                         |
| GET         | /reservations/mine            | ADMIN, MEMBER | 200 OK           | 내 예약 목록 조회                                                                 |
| GET         | /reservations/available-times | PUBLIC        | 200 OK           | 조건부 예약 가능한 시간 목록 조회(예약 날짜, 테마 id)                                          |
| GET         | /reservations/statuses        | ADMIN, MEMBER | 200 OK           | 모든 예약 상태 목록 조회                                                             |
| POST        | /admin/reservations           | ADMIN, MEMBER | 201 CREATED      | 일반 회원의 예약 추가                                                               |
| DELETE      | /admin/reservations/{id}      | ADMIN, MEMBER | 204 NO_CONTENT   | id(PK)에 해당하는 예약 삭제                                                         |
| GET         | /admin/reservations           | ADMIN         | 200 OK           | 모든 회원의 전체 예약 목록 조회                                                         |
| GET         | /admin/reservations/filtered  | ADMIN         | 200 OK           | 조건부 예약 목록 조회 (회원 id, 테마 id, 시작 기간, 끝 기간으로 필터링)                             |

threepebbles replied:
관리자 예약 관리 API와 회원 예약 관리 API를 분리하면서 prefix도 전부 통일해보았습니다!

| HTTP Method | Endpoint                      | AuthRole      | Success          | Description                                                                |
|-------------|-------------------------------|---------------|------------------|----------------------------------------------------------------------------|
| POST        | /reservations                 | ADMIN, MEMBER | 201 CREATED      | 일반 회원의 예약 추가                                                               |
| DELETE      | /reservations/{id}            | ADMIN, MEMBER | 204 NO_CONTENT   | id(PK)에 해당하는 예약 삭제(본인의 예약만 삭제 가능)                                          |
| GET         | /reservations/mine            | ADMIN, MEMBER | 200 OK           | 내 예약 목록 조회                                                                 |
| GET         | /reservations/available-times | PUBLIC        | 200 OK           | 조건부 예약 가능한 시간 목록 조회(예약 날짜, 테마 id)                                          |
| POST        | /admin/reservations           | ADMIN         | 201 CREATED      | 관리자 권한 예약 추가                                                               |
| DELETE      | /admin/reservations/{id}      | ADMIN         | 204 NO_CONTENT   | id(PK)에 해당하는 예약 삭제(모든 예약 삭제 가능)                                            |
| GET         | /admin/reservations           | ADMIN         | 200 OK           | 모든 회원의 전체 예약 목록 조회                                                         |
| GET         | /admin/reservations/filtered  | ADMIN         | 200 OK           | 조건부 예약 목록 조회 (회원 id, 테마 id, 시작 기간, 끝 기간으로 필터링)                             |
| GET         | /admin/reservations/statuses  | ADMIN         | 200 OK           | 모든 예약 상태 목록 조회                                                             |

summerlunaa commented on src/main/java/roomescape/member/infrastructure/MemberRepositoryImpl.java:
```diff
@@ -25,16 +26,19 @@ public void deleteById(final Long id) {
     }
 
     @Override
+    @Transactional(readOnly = true)
```
Transactional을 붙이지 않은 메서드는 어떻게 동작하나요? :)

threepebbles replied:
Transactional 어노테이션을 붙이지 않은 메서드는 기본적으로 auto-commit 모드로 실행됩니다. 각각의 SQL문이 실행된 직후 바로 commit되고, 메서드 진행 도중 예외가 발생해도 이전에 commit된 쿼리에 대해서는 롤백이 불가능합니다.

단, 해당 메서드에 Transactional 어노테이션이 붙어있지 않고, 해당 메서드를 호출한 상위 호출자에 Transactional 어노테이션이 붙어 있는 경우, 호출자에 속하는 트랜잭션으로 묶여서 처리됩니다. 😀

summerlunaa reviewed (changes requested):
헤일러 안녕하세요~
리뷰 빠르게 잘 반영해주셨습니다 👍 제가 드린 질문에도 꼼꼼히 답변 달아주셔서 감사합니다~
마지막으로 코멘트 조금 더 남겼습니다. 확인 부탁드립니다!

threepebbles committed:
refactor: /reservations path prefix 통일

threepebbles committed:
refactor: AdminController 클래스에 RequiresRole 어노테이션을 적용하도록 리팩터링

threepebbles committed:
refactor: ThemeQueryRepository, ThemeCommandRepository 통합

threepebbles committed:
refactor: MemberQueryRepository, MemberCommandRepository 통합

threepebbles committed:
refactor: ReservationTimeQueryRepository, ReservationTimeCommandRepository 통합

threepebbles committed:
refactor: ReservationQueryRepository, ReservationCommandRepository 통합

threepebbles committed:
feat: 관리자 예약 서비스와 회원 예약 서비스 분리, 관리자는 과거 시간에 예약이 가능하도록 기능 수정

threepebbles committed:
refactor: 데이터 존재 여부 확인에 findById 대신 ExistsById 사용

threepebbles committed:
fix: 삭제하려는 행이 존재하지 않는 경우, 커스텀 예외를 발생시키도록 수정

threepebbles committed:
test: 관리자 예약 관리 서비스 테스트 추가, 회원 예약 관리 서비스 테스트 보강

threepebbles committed:
test: 테스트 픽스처 네이밍 컨벤션 수정

threepebbles committed:
chore: 패키지 이동

threepebbles committed:
test(ThemeService): 인기 테마 목록 조회 기능 테스트 추가

threepebbles committed:
test(ReservationRepository): 테마, 회원, 날짜 간격 기준 예약 조회 기능 테스트 추가

threepebbles committed:
test(AdminReservationService): 테마, 회원, 날짜 간격 기준 예약 조회 기능 테스트 개선

threepebbles committed:
docs: 관리자 예약 관리 API와 회원 예약 관리 API 분리에 따라 변경된 API 스펙 반영

threepebbles committed:
refactor: 예약 상태 전체 목록 조회 기능를 관리자 전용 기능으로 수정

threepebbles committed:
docs: 변경된 API 스펙 반영

threepebbles committed:
docs: 중복 행 삭제

threepebbles commented:
안녕하세요 파랑! 피드백 주신 내용 다시 반영해보았습니다
리뷰 잘 부탁드립니다 😄

summerlunaa reviewed (approved):
헤일러 안녕하세요!
이번에도 리뷰 반영 잘 해주셨습니다. 여러 고민이 많았을 것 같은데 잘 정리해서 의견 남겨주셨네요 👍 
이제 다음 단계로 넘어가도 좋을 것 같아요 :) 3-4단계도 파이팅입니다 💪
